### PR TITLE
[SHELLEXT][FONTEXT] An attempt to implement IDropTarget

### PR DESCRIPTION
**Pull request by katahiromz**
## Purpose

JIRA issue: [CORE-12861](https://jira.reactos.org/browse/CORE-12861)

### CODE REVIEW

File: dll/shellext/fontext/CFontExt.cpp [hunk 1]
```diff
@@ -363,3 +365,176 @@ STDMETHODIMP CFontExt::GetClassID(CLSID *lpClassId)
...
 }
 
+// *** IDropTarget methods ***
+STDMETHODIMP CFontExt::DragEnter(IDataObject* pDataObj, DWORD grfKeyState, POINTL pt, DWORD* pdwEffect)
+{
+    FORMATETC formatETC;
+    HRESULT hr = DoGetFormatETC(pDataObj, formatETC);
+    if (hr != S_OK)
+    {
+        *pdwEffect = DROPEFFECT_NONE;
```
Comment (learn-more) [inline dll/shellext/fontext/CFontExt.cpp]: is `pdwEffect` always a required parameter?

File: dll/shellext/fontext/CFontExt.cpp [hunk 2]
```diff
@@ -363,3 +365,176 @@ STDMETHODIMP CFontExt::GetClassID(CLSID *lpClassId)
...
+    return S_OK;
+}
+
+STDMETHODIMP CFontExt::Drop(IDataObject* pDataObj, DWORD grfKeyState, POINTL pt, DWORD* pdwEffect)
+{
+    FORMATETC formatETC;
+    HRESULT hr = DoGetFormatETC(pDataObj, formatETC);
+    if (hr != S_OK)
+    {
+        *pdwEffect = DROPEFFECT_NONE;
```
Comment (learn-more) [inline dll/shellext/fontext/CFontExt.cpp]: Same here, should be checked first maybe?

File: dll/shellext/fontext/CFontExt.cpp [hunk 3]
```diff
@@ -363,3 +365,176 @@ STDMETHODIMP CFontExt::GetClassID(CLSID *lpClassId)
...
+    hr = pDataObj->GetData(&formatETC, &medium);
+    if (FAILED(hr))
+    {
+        *pdwEffect = DROPEFFECT_NONE;
+        return DRAGDROP_S_CANCEL;
+    }
+
+    BOOL bOK = TRUE;
+    CAtlArray<CStringW> FontPaths;
+    HDROP hDrop = (HDROP)GlobalLock(medium.hGlobal);
```
Comment (learn-more) [inline dll/shellext/fontext/CFontExt.cpp]: Better to use HIDA here, see helper functions `HIDA_GetPIDLFolder` and `HIDA_GetPIDLItem`

File: dll/shellext/fontext/CFontExt.cpp [hunk 4]
```diff
@@ -363,3 +365,176 @@ STDMETHODIMP CFontExt::GetClassID(CLSID *lpClassId)
...
+
+    if (!bOK)
+        return E_FAIL;
+
+    WCHAR szFontsDir[MAX_PATH];
+    GetWindowsDirectoryW(szFontsDir, _countof(szFontsDir));
+    PathAppendW(szFontsDir, L"Fonts");
+
+    HKEY hkeyFonts = NULL;
+    RegOpenKeyExW(HKEY_LOCAL_MACHINE,
```
Comment (learn-more) [inline dll/shellext/fontext/CFontExt.cpp]: `CFontCache::Read` uses same paths, maybe they can be shared?

File: dll/shellext/fontext/CFontExt.cpp [hunk 5]
```diff
@@ -363,3 +365,176 @@ STDMETHODIMP CFontExt::GetClassID(CLSID *lpClassId)
...
+        GlobalUnlock(medium.hGlobal);
+    }
+    ReleaseStgMedium(&medium);
+
+    if (!bOK)
+        return E_FAIL;
+
+    WCHAR szFontsDir[MAX_PATH];
+    GetWindowsDirectoryW(szFontsDir, _countof(szFontsDir));
+    PathAppendW(szFontsDir, L"Fonts");
```
Comment (learn-more) [inline dll/shellext/fontext/CFontExt.cpp]: Other places in this library `SHGetFolderPathW(NULL, CSIDL_FONTS, NULL, 0, Path)` is used

File: dll/shellext/fontext/CFontExt.hpp [hunk 6]
```diff
@@ -81,4 +88,8 @@ class CFontExt :
         COM_INTERFACE_ENTRY_IID(IID_IPersistFolder, IPersistFolder)
```
Comment (learn-more) [inline dll/shellext/fontext/CFontExt.hpp]: If `IDropTarget` should also be available via `QueryInterface`, add it in this table.

File: dll/shellext/fontext/CFontExt.cpp [hunk 7]
```diff
@@ -363,3 +365,176 @@ STDMETHODIMP CFontExt::GetClassID(CLSID *lpClassId)
...
+        return E_FAIL;
+
+    RemoveFontResourceW(pszFileTitle);
+
+    StringCchCopyW(szDestFile, sizeof(szDestFile), pszFontsDir);
+    PathAppendW(szDestFile, pszFileTitle);
+    if (!CopyFileW(pszFontPath, szDestFile, FALSE))
+        return E_FAIL;
+
+    if (!AddFontResourceW(pszFileTitle))
```
Comment (learn-more) [inline dll/shellext/fontext/CFontExt.cpp]: Why is this needed?

File: dll/shellext/fontext/CFontExt.cpp [hunk 8]
```diff
@@ -363,3 +365,176 @@ STDMETHODIMP CFontExt::GetClassID(CLSID *lpClassId)
...
+            DragQueryFile(hDrop, iFile, szPath, _countof(szPath));
+
+            if (PathIsDirectoryW(szPath))
+            {
+                bOK = FALSE;
+                break;
+            }
+
+            LPCWSTR pchDotExt = PathFindExtensionW(szPath);
+            if (_wcsicmp(pchDotExt, L".ttf") != 0 &&
```
Comment (learn-more) [inline dll/shellext/fontext/CFontExt.cpp]: Are these the only extensions supported?
Can we add this in a list somewhere, so that we can just loop that list here?

File: dll/shellext/fontext/CFontExt.cpp [hunk 9]
```diff
@@ -363,3 +365,176 @@ STDMETHODIMP CFontExt::GetClassID(CLSID *lpClassId)
...
+                  0,
+                  KEY_READ | KEY_WRITE,
+                  &hkeyFonts);
+    if (!hkeyFonts)
+        return E_FAIL;
+
+    for (size_t iItem = 0; iItem < FontPaths.GetCount(); ++iItem)
+    {
+        HRESULT hr = DoInstallFontFile(FontPaths[iItem], szFontsDir, hkeyFonts);
+        if (FAILED(hr))
```
Comment (learn-more) [inline dll/shellext/fontext/CFontExt.cpp]: `FAILED_UNEXPECTEDLY`

File: dll/shellext/fontext/CFontExt.cpp [hunk 10]
```diff
@@ -363,3 +365,176 @@ STDMETHODIMP CFontExt::GetClassID(CLSID *lpClassId)
...
+    WCHAR szFontName[512];
+    if (!DoGetFontTitle(pszFontPath, szFontName))
+        return E_FAIL;
+
+    RemoveFontResourceW(pszFileTitle);
+
+    StringCchCopyW(szDestFile, sizeof(szDestFile), pszFontsDir);
+    PathAppendW(szDestFile, pszFileTitle);
+    if (!CopyFileW(pszFontPath, szDestFile, FALSE))
+        return E_FAIL;
```
Comment (learn-more) [inline dll/shellext/fontext/CFontExt.cpp]: Logging here would be nice, since this is unexpected.

File: dll/shellext/fontext/CFontExt.cpp [hunk 11]
```diff
@@ -363,3 +365,176 @@ STDMETHODIMP CFontExt::GetClassID(CLSID *lpClassId)
...
+
+    if (!AddFontResourceW(pszFileTitle))
+        return E_FAIL;
+
+    DWORD cbData = (wcslen(pszFileTitle) + 1) * sizeof(WCHAR);
+    LONG nError = RegSetValueExW(hkeyFonts, szFontName, 0, REG_SZ, (const BYTE *)szFontName, cbData);
+    if (nError)
+    {
+        return E_FAIL;
```
Comment (learn-more) [inline dll/shellext/fontext/CFontExt.cpp]: Do we need to delete the file from the fonts folder?

File: dll/shellext/fontext/CFontExt.cpp [hunk 12]
```diff
@@ -363,3 +365,176 @@ STDMETHODIMP CFontExt::GetClassID(CLSID *lpClassId)
...
+    }
+
+    return E_FAIL;
```
Comment (learn-more) [inline dll/shellext/fontext/CFontExt.cpp]: S_OK ?

File: dll/shellext/fontext/CFontExt.cpp (hunk 1 above)
Comment (katahiromz) [reply dll/shellext/fontext/CFontExt.cpp]: https://docs.microsoft.com/en-us/windows/win32/api/oleidl/nf-oleidl-idroptarget-dragenter
>On return, the method must write the effect, one of the DROPEFFECT flags, to the pdwEffect parameter.

File: dll/shellext/fontext/CFontExt.cpp (hunk 2 above)
Comment (katahiromz) [reply dll/shellext/fontext/CFontExt.cpp]: OK.

File: dll/shellext/fontext/CFontExt.cpp (hunk 5 above)
Comment (katahiromz) [reply dll/shellext/fontext/CFontExt.cpp]: OK.

File: dll/shellext/fontext/CFontExt.cpp (hunk 9 above)
Comment (katahiromz) [reply dll/shellext/fontext/CFontExt.cpp]: OK.

File: dll/shellext/fontext/CFontExt.cpp (hunk 3 above)
Comment (katahiromz) [reply dll/shellext/fontext/CFontExt.cpp]: OK.

File: dll/shellext/fontext/CFontExt.cpp (hunk 8 above)
Comment (katahiromz) [reply dll/shellext/fontext/CFontExt.cpp]: `IsFontDotExt` function will cover it.

File: dll/shellext/fontext/CFontExt.cpp (hunk 4 above)
Comment (katahiromz) [reply dll/shellext/fontext/CFontExt.cpp]: font cache is not relevant to this code.

File: dll/shellext/fontext/CFontExt.hpp (hunk 6 above)
Comment (katahiromz) [reply dll/shellext/fontext/CFontExt.hpp]: OK.

File: dll/shellext/fontext/CFontExt.cpp (hunk 7 above)
Comment (katahiromz) [reply dll/shellext/fontext/CFontExt.cpp]: https://docs.microsoft.com/en-us/windows/win32/api/wingdi/nf-wingdi-addfontresourcew
>The AddFontResource function adds the font resource from the specified file to the system font table. The font can subsequently be used for text output by any application.

`CopyFileW` just copys the file to the filesystem. `CopyFileW` won't install the font file.

Comment (katahiromz) [reply dll/shellext/fontext/CFontExt.cpp]: Font engine became registry-based, so simply copying doesn't install font.

File: dll/shellext/fontext/CFontExt.cpp [hunk 13]
```diff
@@ -363,3 +365,161 @@ STDMETHODIMP CFontExt::GetClassID(CLSID *lpClassId)
...
+
+    WCHAR szRoot[MAX_PATH];
+    PCUIDLIST_ABSOLUTE pidlRoot = HIDA_GetPIDLFolder(cida);
+    SHGetPathFromIDListW(pidlRoot, szRoot);
+
+    BOOL bOK = TRUE;
+    CAtlArray<CStringW> FontPaths;
+    for (UINT n = 0; n < cida->cidl; ++n)
+    {
+        const FontPidlEntry* fontEntry = _FontFromIL(HIDA_GetPIDLItem(cida, n));
```
Comment (learn-more) [inline dll/shellext/fontext/CFontExt.cpp]: The pidl's do not have to come from the fontext itself, so this might not work.
In this case you'd have to bind a shellfolder to the pidlRoot, and ask that for it's child.

File: dll/shellext/fontext/CFontExt.cpp [hunk 14]
```diff
@@ -363,3 +365,161 @@ STDMETHODIMP CFontExt::GetClassID(CLSID *lpClassId)
...
+            break;
+        }
+
+        FontPaths.Add(File);
+    }
+
+    if (!bOK)
+        return E_FAIL;
+
+    HKEY hkeyFonts = NULL;
```
Comment (learn-more) [inline dll/shellext/fontext/CFontExt.cpp]: If you use CRegKey, you can simply return failure codes in the loop below

File: dll/shellext/fontext/CFontExt.cpp (hunk 7 above)
Comment (learn-more) [reply dll/shellext/fontext/CFontExt.cpp]: Ah yes, this makes it available immediately to the applications without a reboot.

File: dll/shellext/fontext/CFontExt.cpp (hunk 14 above)
Comment (katahiromz) [reply dll/shellext/fontext/CFontExt.cpp]: OK.

File: dll/shellext/fontext/CFontExt.cpp (hunk 13 above)
Comment (katahiromz) [reply dll/shellext/fontext/CFontExt.cpp]: I used `ILCombine`.

File: dll/shellext/fontext/CFontExt.cpp (hunk 4 above)
Comment (katahiromz) [reply dll/shellext/fontext/CFontExt.cpp]: Sorry, I have no idea. What can I do?

**Comment by katahiromz** [comment]
To go, or not to go?

**Comment by learn-more** [comment]
It does not have to be finished, if it is an improvement over the previous state I'd say go for it.
